### ISSUE TITLE
Separate Story XML files bucket from the Material bucket

### ISSUE THREAD

**Issue by jetkan-yk** [source-academy/cadet#592]
Currently the Game Dev page is sharing the same AWS bucket with the Materials page.

![image](https://user-images.githubusercontent.com/36299141/82407776-a3afbb00-9a9c-11ea-8ae1-3db6a2f4a405.png)

Consider only display Test Mission folder in Materials page, and Test Stories folder in Game Dev page.

Or alternatively, create a new Game Dev bucket in AWS.

**Comment by martin-henz** [comment]
This issue is not clear. You say "consider only display": Is this a display issue or a structural issue?

What are these folders "Test Stories" and "Test Missions" in the window?

Can you describe the current state of affairs? Do we have a wiki page on this? 
* Where are the "Materials" files stored?
* Where are the assessment files (missions, quests) stored?
* Where are the story XML files stored?

What changes to this state of affairs do you propose?

**Comment by angelsl** [comment]
> Currently the Game Dev page is sharing the same AWS bucket with the Materials page.

Actually this is because the frontend Game Dev component seems to be just using the Materials API endpoints, which is kind of weird: https://github.com/source-academy/cadet-frontend/pull/1058/files#diff-8c991eec9f049b551867f4f6803c3ac2

I can make separate endpoints for this. But if we're going to retire the student-facing Materials part, then maybe I'll just rename the endpoints to be "gamedev" or something like that.

Also, to answer this:

> * Where are the "Materials" files stored?

In an S3 bucket `prod-cadet-materials` for production, and `stg-cadet-materials` for staging. These buckets are uploaded to via the backend, using the "Upload" function in the frontend.

> * Where are the assessment files (missions, quests) stored?

The data that is actually served to students is stored in the PostgreSQL database, which is imported from the XMLs in the Bitbucket repositories periodically. There are no S3 buckets involved here.

The Ground Control feature lets staff/admin roles upload XML which is imported directly into the PostgreSQL database. (Not sure if this is a good idea... because then Bitbucket becomes no longer the single source of truth. But that's what it is currently.)

> * Where are the story XML files stored?

They are currently stored in an S3 bucket called `source-academy-assets`.

Doesn't seem like the Game Dev component added a way to actually upload to this bucket, which is strange. (It still has to be done manually.)

**Comment by angelsl** [comment]
So maybe what should happen is, this Game Dev component, which currently has a "Test" mode and a "Live" mode, should allow the game devs to view/manage the contents of `source-academy-assets` for "Live" mode, and some other S3 bucket for "Test" mode.

That might have been the intention of the Game Dev team. Not sure why it ended up going into Materials..

**Comment by martin-henz** [comment]
> So maybe what should happen is, this Game Dev component, which currently has a "Test" mode and a "Live" mode, should allow the game devs to view/manage the contents of `source-academy-assets` for "Live" mode, and some other S3 bucket for "Test" mode.
> 
> That might have been the intention of the Game Dev team. Not sure why it ended up going into Materials..

I advised the team to re-use "Materials", to keep the system simple.

**Comment by angelsl** [comment]
> * Live mode: the developer experiences whatever the student experiences: all published stories
> * Test mode: the developer experiences all stories: published or unpublished.
> 
> 
> Is that your proposal @angelsl ?

Not exactly my proposal; it already sort-of exists in the game dev component as far as I can tell: https://github.com/source-academy/cadet-frontend/pull/1058/files#diff-44010d88d325504868a90b2b7497ae73

**Comment by martin-henz** [comment]
(I think I accidentally closed this issue.)

**Comment by martin-henz** [comment]
To streamline all of this, why don't we store the stories in the database as well? Then we can enjoy the parsing and error detection infrastructure that we already have for the assessment XML files.

Regarding: "(Not sure if this is a good idea... because then Bitbucket becomes no longer the single source of truth. But that's what it is currently.)"

This was a deliberate decision. I don't think it was a good idea to tie the Source Academy to bitbucket in the way we did it before.

So we cut that dependency to simplify the overall system. The authors of missions (and stories) are free to manage these files in whatever way they like, and upload them to the backend when they are ready.

**Comment by angelsl** [comment]
> To streamline all of this, why don't we store the stories in the database as well? Then we can enjoy the parsing and error detection infrastructure that we already have for the assessment XML files.

That's possible.

Though, the backend doesn't currently handle stories at all. The frontend loads it directly from the S3 bucket.

**Comment by martin-henz** [comment]
I'm arguing for simplicity here. It's odd to keep some XML files in S3 buckets and others in the database, especially when they are so closely related.

**Comment by jetkan-yk** [comment]
There is a discussion happening concurrently regarding the frontend of Ground Control and Game Dev. https://github.com/source-academy/cadet-frontend/issues/1131

TLDR: I'm proposing that we store **ALL** published/unpublished Assessment & Story XMLs in **ONE** PostgreSQL database. The Ground Control is the only gateway to modify the XML files (open dates, close dates, etc. and this will deprecate the dates attributes in the Assessment XML tags).

The publish button also remove the need of having a 'live' database and 'test' database. If an XML is published, staffs and students have access to it (e.g. play in game or view from the Assessment tabs). If an XML is unpublished, only staffs have access.

**Comment by jetkan-yk** [comment]
The decisions for this issue is documented in this wiki page: https://github.com/source-academy/cadet-frontend/wiki/Design-Decision:-Ground-Control,-Assessment-&-Story-XML-database,-Game-Storyline
